app/helpers/audits_helper.rb

```diff
@@ -244,4 +241,102 @@ def key_to_class(key, audit)
     auditable_type = (audit.auditable_type == 'Host::Base') ? 'Host::Managed' : audit.auditable_type
     auditable_type.constantize.reflect_on_association(key.sub(/_id(s?)$/, '\1'))&.klass
   end
+
+  def rebuild_audit_changes(audit)
+    css_class_name = audit.action.eql?('destroy') ? 'show-old' : 'show-new'
+    audit.audited_changes.map do |name, change|
+      next if change.nil? || change.to_s.empty?
+      next if name == 'template'
+      rec = { :name => name.humanize }
+      if audit.action == 'update'
+        rec[:change] = change.map.with_index do |v, i|
+          change_info_hash(name, v, (i == 0) ? 'show-old' : 'show-new')
+        end
+      else
+        rec[:change] = (rec[:change] || []).push(change_info_hash(name, change, css_class_name))
+      end
+      rec
+    end.compact
+  end
+
+  def change_info_hash(name, change, css_class = 'show-new')
+    { :css_class => css_class, :id_to_label => id_to_label(name, change, truncate: false) }
+  end
+
+  def fetch_affected_locations(audit)
+    base = audit.locations.authorized(:view_locations)
+    return [] if base.empty?
+
+    authorizer = Authorizer.new(User.current, base)
+    base.map do |location|
+      options = hash_for_edit_location_path(location).merge(:auth_object => location, :permission => 'edit_locations', :authorizer => authorizer)
+      construct_options(location.name, edit_location_path(location), options)
+    end
+  end
+
+  def fetch_affected_organizations(audit)
+    base = audit.organizations.authorized(:view_organizations)
+    return [] if base.empty?
+
+    authorizer = Authorizer.new(User.current, base)
+    base.map do |organization|
+      options = hash_for_edit_organization_path(organization).merge(:auth_object => organization, :permission => 'edit_organizations', :authorizer => authorizer)
+      construct_options(organization.name, edit_organization_path(organization), options)
+    end
+  end
+
+  def construct_options(affected_obj_name, affected_obj_url, options = {})
+    if authorized_for(options)
+      {'name' => affected_obj_name, 'url' => affected_obj_url}
+    else
+      {'name' => affected_obj_name, 'url' => '#', 'css_class' => "disabled", 'disabled' => true}
+    end
+  end
+
+  def time_of_creation(audit)
+    time = audit.created_at
+    return { 'value' => _('N/A') } if time.nil?
+    { 'title' => date_time_relative_value(time),
+      'value' => date_time_absolute_value(time, :short) }
+  end
+
+  def audit_title_url(audit)
+    keytype_array = Audit.find_complete_keytype_array(audit.auditable_type)
+    filter = "type = #{keytype_array.first} and auditable_id = #{audit.auditable_id}" if keytype_array.present?
+    (filter ? audits_path(:search => filter) : nil)
+  end
+
+  def user_info(audit)
+    return {} if audit.username.nil?
+    login = audit.user.login rescue nil # aliasing the user method sometimes yields strings
```

Comment:
this and the same line in `audit_user` could be replaced by `audit.user_login` if we add  `delegate :login, :to => :user, :allow_nil => true, :prefix => true`  to audit class